Role CRUD + 모든 Roles 조회 기능 생성

DeiLee0913 opened the pull request:
## 📝 작업 내용

- [x]  역할 생성 API
- [x]  ID로 역할 조회 API
- [x]  NAME으로 역할 조회 API
- [x]  NAME으로 역할 수정 API
- [x]  ID로 역할 삭제 API
- [x]  NAME으로 역할 삭제 API
- [x]  프로필 ID로 모든 역할 조회 API
- [x]  동일 프로필 내 중복 역할 검증
- [x]  역할에 대한 접근 권한 검증
- [x]  roleName으로 role에 접근 시 현재 유저의 profile에서만 조회

## 📸 결과

- ‘POST’ createRole(/api/v1/role)
    - RequestHeader
        - Authorization
    - RequestBody
        
        ```json
        {
          "roleName": "Family",
          "securityLevel": "LEVEL_THREE"
        }
        ```
        
    - ReponseBody
        
        ```json
        {
            "roleId": 8,
            "profileId": 1,
            "roleName": "Friend",
            "securityLevel": "LEVEL_FIVE"
        }
        ```
        
- ‘GET’ getRoleById(/api/v1/role/{roleId})
    - RequestHeader
        - Authorization
    - ResponseBody
        
        ```json
        {
            "roleId": 8,
            "profileId": 1,
            "roleName": "Friend",
            "securityLevel": "LEVEL_FIVE"
        }
        ```
        
- ‘GET’ getRoleByName(/api/v1/role/by-name)
    - RequestHeader
        - Authorization
    - RequestBody
        
        ```java
        {
            "roleName": "Friend"
        }
        ```
        
    - RespnseBody
        
        ```json
        {
            "roleId": 8,
            "profileId": 1,
            "roleName": "Friend",
            "securityLevel": "LEVEL_FIVE"
        }
        ```
        
- ‘GET’ getAllRoles(/api/v1/role/profile)
    - RequestHeader
        - Authorization
    - ResponseBody
        
        ```json
        [
            {
                "roleName": "Family",
                "securityLevel": "LEVEL_TWO"
            },
            {
                "roleName": "Lover",
                "securityLevel": "LEVEL_THREE"
            },
            {
                "roleName": "BF",
                "securityLevel": "LEVEL_FOUR"
            },
            {
                "roleName": "Friend",
                "securityLevel": "LEVEL_FIVE"
            }
        ]
        ```
        
- ‘PUT’ updateRoleById(/api/v1/role/{roleId})
    - RequestHeader
        - Authorization
    - RequestBody
        
        ```json
        {
          "roleName": "Friends",
          "securityLevel": "LEVEL_FOUR"
        }
        ```
        
    - RespnseBody
        
        ```json
        {
            "roleId": 8,
            "profileId": 1,
            "roleName": "Friends",
            "securityLevel": "LEVEL_FOUR"
        }
        ```
        
- ‘PUT’ updateRoleByName(/api/v1/role/by-name)
    - RequestHeader
        - Authorization
    - RequestBody
        
        ```json
        {
            "originalName": "Friends",
            "newName": "Friend",
            "newSecurityLevel": "LEVEL_FIVE"
        }
        ```
        
    - RespnseBody
        
        ```json
        {
            "roleId": 8,
            "profileId": 1,
            "roleName": "Friend",
            "securityLevel": "LEVEL_FIVE"
        }
        ```
        
- ‘DELETE’ deleteRoleById(/api/v1/role/{roleId})
    - RequestHeader
        - Authorization
    - ResponseBody
        
        ```java
        역할이 삭제되었습니다.
        ```
        
- ‘DELETE’ deleteRoleByName(/api/v1/role/by-name)
    - RequestHeader
        - Authorization
    - RequestBody
        
        ```json
        {
            "roleName": "지인"
        }
        ```
        
    - ResponseBody
        
        ```json
        역할이 삭제되었습니다.
        ```
        

## 💬 리뷰 요구사항 (선택)

## #️⃣ 연관된 이슈 (선택)

DeiLee0913 committed:
✨ feat: Role Dto 생성, Domain 업데이트

SecurityLevel 필드 값을 String에서 Int로 변경

DeiLee0913 committed:
✨ feat: Role CRUD + 모든 Role 받기 기능 구현

chlqls commented on src/main/java/jusomejusome/togather/role/dto/request/RoleCreateReqDto.java:
```diff
@@ -0,0 +1,18 @@
+    private Integer securityLevel;
+
+    @Builder
+    public RoleCreateReqDto(long profileId, String roleName, int securityLevel) {
```
큰 문제는 아닌데 혹시 모르니 여기도 데이터타입 통일하는게 좋을 것 같아요-! (`long` -> `Long`, `int` -> `Integer`)

DeiLee0913 replied:
createReqDto -> reqDto로 수정하면서 profileId는 삭제했고 -> user로 토큰으로 profileId 받아서 사용, securityLevel은 SecurityLevel 타입으로 통일했습니다 :)

chlqls commented on src/main/java/jusomejusome/togather/role/dto/request/RoleUpdateReqDto.java:
```diff
@@ -0,0 +1,16 @@
+import lombok.Getter;
+
+@Getter
+public class RoleUpdateReqDto {
```
`securityLevel`도 변경가능하니까 여기서 같이 시큐리티레벨 보내서 같이 업데이트하면 좋을 것 같아용

DeiLee0913 replied:
작업 완료했습니다 :)

chlqls commented on src/main/java/jusomejusome/togather/role/service/RoleService.java:
```diff
@@ -0,0 +1,116 @@
+    public List<String> getAllRoleNames(Long profileId) {
+        List<Role> roles = roleRepository.findByProfile_profileId(profileId);
+        if(roles.isEmpty()) {
+            throw new CustomException(ErrorCode.PROFILE_NOT_FOUND_EXCEPTION);
```
여기 `PROFILE_NOT_FOUND_EXCEPTION` 말고 새로운 에러코드로 예외처리 해야할 것 같아여!

DeiLee0913 replied:
profileId를 dto에서 말고 user에서 token을 통해 받아오면서 관련 내용 삭제됐습니다 :)

chlqls commented on src/main/java/jusomejusome/togather/role/service/RoleService.java:
```diff
@@ -0,0 +1,116 @@
+    @Transactional
+    public RoleResDto updateRoleByName(RoleUpdateReqDto updateDto) {
+        // 동일한 이름을 가진 카테고리가 있는지 확인
+        if(roleRepository.existsByRoleName(updateDto.getNewName())) {
```
카테고리랑 마찬가지로, 동일한 이름의 role이어도 profileId가 다를 수 있으니까 `existsByRoleNameAndProfile_ProfileId`처럼 profileId도 같이 확인해주면 좋을 것 같습니다~

DeiLee0913 replied:
profileId도 함께 확인해서 조회하도록 작업 완료했습니다 :)

chlqls commented on src/main/java/jusomejusome/togather/role/controller/RoleController.java:
```diff
@@ -0,0 +1,68 @@
+        return ResponseEntity.ok(role);
+    }
+
+    @GetMapping("/by-name/{roleName}")
```
여기서 roleName이 한글이면 에러나는 것 같아요! 아래 업데이트도 마찬가지로,,,, 한번 확인 부탁드려욤

DeiLee0913 replied:
@PathVariable로는 id만 받고 String은 Dto로 받아오도록 작업 완료했습니다 :)

DeiLee0913 committed:
🌟update: Role Domain, Dto 생성

- securityLevel enum 변경사항 반영(타브랜치에서 수정됨)해 Role domain 변경
- RoleCreateReqDto -> ReqDto로 이름 변경하고 profileId 필드 삭제

DeiLee0913 committed:
✨ feat: RoleNameReqDto & RoleInfoDto 생성

DeiLee0913 committed:
🌟update: 동일 프로필 내 중복 역할 검증 & 역할에 대한 접근 권한 검증

- getAllRoles 반환형을 List<String>에서 List<RoleInfoDto>로 변경해 RoleName과 SecurityLevel을 같이 전달
- roleName으로 role에 접근할 때 현재 유저 정보로 profileId 받아와 profile 내에서 role을 찾도록 함

DeiLee0913 committed:
🌟update: dev와 충돌 해결 위해 errorCode 추가

DeiLee0913 committed:
🌟update: dev와 충돌 해결 위해 errorCode 수정

DeiLee0913 committed:
Merge branch 'dev' into feature/role
